### PR TITLE
Integration: marian new on-the-fly-decoder for bergamot

### PR DESCRIPTION
**Pull request by jerinphilip**
WIP repurpose of MTS to translate on the fly and provide a consumable API for bergamot-translator.
from [browsermt/mts](https://github.com/browsermt/mts/tree/nuke).

### CODE REVIEW

File: 3rd_party/CMakeLists.txt
Comment (jerinphilip) [reply 3rd_party/CMakeLists.txt]: The latest commit reverts @abhi-agg's changes to `CMakeLists.txt` to add compilation flags, while enabling CI for MacOS and Ubuntu triggered on pull requests/pushes to the main branch. @abhi-agg: Please ensure that if you are modifying code (CMakeLists), it does not affect Ubuntu builds.

Comment (abhi-agg) [reply 3rd_party/CMakeLists.txt]: @jerinphilip I checked the failure logs that you provided and it is coming from the sources of bergamot translator repo and not from marian-dev sources.

Importing Compilation Flags of marian and applying them to the whole bergamot translator sources doesn't make sense as lot of them (e.g `-msse2 -msse3 -msse4.1 -msse4.2 -mavx -mavx2 -mavx512f -DUSE_SENTENCEPIECE -DMKL_ILP64`) are not even relevant for these sources. A better way would be to explicitly specifying flags that are relevant for optimization (e.g.` -O3`) in this project.

I will provide a fix for this.

Comment (jerinphilip) [reply 3rd_party/CMakeLists.txt]: This is not blocking integration, which is pressing at this stage. I'm happy with a working build at the moment, the apt CMake configurations can come later. Moving this to issue #14 therefore, preferably to be taken outside this PR, therefore.

**Comment by abhi-agg** [comment]
@jerinphilip Is there anything else that is blocking this PR from merging?

**Comment by jerinphilip** [comment]
@abhi-agg: Connecting your API with marian::bergamot::Service was a requirement before merge in my mind.

 https://github.com/browsermt/bergamot-translator/pull/8/commits/08a7358c3d6caf55a6eb38f24b9955f474cd9729 adds a working draft completing the said requirement. Probably needs some finishing touches. 
 
 `TranslationRequest` and associated functionality are not going to be complete in this one, I hope we're in agreement to move it outside this PR.

File: app/main.cpp
Comment (abhi-agg) [inline app/main.cpp:23]: Please remove `std::cout << config << std::endl;` from here.

**Comment by abhi-agg** [comment]
@jerinphilip From this point inwards, could you please do modular changes in this PR?

e.g. There could have been 1 commit to replace `TranslationModelConfiguration` with `std::string` in the API. Btw, there is some amount of cleanup left to be done pertaining to this change. That is why, I created this PR https://github.com/browsermt/bergamot-translator/pull/15.

Another commit to connect TranslationModel to Service class.

It becomes easy to review as well. Creating Modular changes is a part of good git practices :)

**Comment by kpu** [comment]
I agree with what @abhi-agg says: git commits should correspond to specific changes.

**Comment by abhi-agg** [comment]
I pushed the commit that does all the left over work for `TranslationModelConfiguration` to `std::string` migration in APIs.

File: src/translator/translation_result.h
Comment (kpu) [inline src/translator/translation_result.h:48]: Why did these become public?

**Comment by kpu** [comment]
What is this I hear about two translation results, one @abhi-agg and one @jerinphilip?  This is silly.  There should only be one.

**Review by kpu — CHANGES_REQUESTED** [review]

**Comment by kpu** [comment]
Remove `Histories` from @jerinphilip translation result.  He was unable to coherently explain why it is there.  Something something OutputCollector, still unclear on this.

**Comment by jerinphilip** [comment]
> What is this I hear about two translation results, one @abhi-agg and one @jerinphilip? This is silly. There should only be one. This will mean changes from both of you.

@kpu: Trying to add some coherance.

```bash
marian-decoder = Translator + OutputCollector
marian-decoder-new = Service + OutputCollector
```

**Move `Service` to marian-dev and benchmark with marian-decoder**: To easily benchmark `marian-decoder` vs `marian-decoder-new`, I need access to `Histories`, which I can use in conjunction with `OutputCollector` and additionally use all the parsing (options) etc code that comes for free with `OutputCollector`. Note that the OutputCollector even uses [Alignment](https://github.com/browsermt/marian-dev/blob/c10b5a248d9d511ddd01e12101cded3b45dbac62/src/translator/output_printer.cpp#L7), which needs to be constructed from History-> nbest -> alignment. 
* I want to add this benchmark script and a regression-test eventually to ensure new commits also do not break speed. Do we retain the benchmarking requirement? The [benchmark script](https://github.com/browsermt/mts/blob/nuke/scripts/benchmark-md-bd.sh) as it is written now is enabled by this adjustment. 
* `Service` is being written as a replacement for `Translator`, based on what I understand from meetings (Also linking: `Service` as solution for https://github.com/marian-nmt/marian-dev/issues/775). I think in the ideal case, `Service` should return a `marian::bergamot::TranslationResult` or a `marian::bergamot::Response = f(marian::bergamot::Request)`, instead of directly connecting to `UnifiedAPI::TranslationResult`. This allows `Service` to be plucked from here and added to marian-dev. We can probably rename `marian::begamot::TranslationResult` if that solves the ugliness. If we have to do the moving with `UnifiedAPI::TranslationResult`, we make `Service` dependent on bergamot-translator.

If @kpu thinks just an `std::cout` of outputs are all that suffice to replace `OutputCollector` and benchmark, I can do that, I'll write that one up without `Histories`.  But I get a lot of things for free if I get to keep `Histories` in `marian::bergamot::TranslationResult` (including easy prototyping of Alignments in the future). 

`Histories` is not allowed by design in `UnifiedAPI::TranslationResult`, as it is supposed to be agnostic to the `TranslationModel` implementation, which is why there are two `TranslationResult`s and an extra conversion for now.

File: app/main.cpp
Comment (jerinphilip) [reply app/main.cpp]: Covered in https://github.com/browsermt/bergamot-translator/pull/8/commits/e76a602dc7205567fdcb76820349bafa8f51bf51. Also removes a file-path in output, enabling a test for bergamot-translator-app in [browsermt/bergamot-translator-tests](https://github.com/browsermt/bergamot-translator-tests).

File: src/translator/sanelogging.h
Comment (kpu) [inline src/translator/sanelogging.h:9]: What is PLOG?  You said per-worker logging.  Can I have a comment to that effect?

Comment (kpu) [reply src/translator/sanelogging.h:9]: Or you could remove if not needed anymore.  Or comment that it is for debugging only.

**Comment by jerinphilip** [comment]
@abhi-agg wants main updated as single source of truth. Before merge, it's agreed that we'll get CI, which is failing above to build.

For this, three variants of workflows are to be added (x 2 for Ubuntu, Mac)

- [ ] native build with vanilla marian (no WASM modifications, should build against all marian-capabilities) @jerinphilip 
- [ ] native build with marian-modifications for WASM @abhi-agg 
- [ ] wasm build with marian-modifications for WASM @abhi-agg

**Comment by jerinphilip** [comment]
@abhi-agg http://vali.inf.ed.ac.uk/jenkins/job/bergamot-translator/33/console
```
../../../libmarian.a(graph_group_sync.cpp.o): In function `operator()':
/var/lib/jenkins/workspace/bergamot-translator/3rd_party/marian-dev/src/training/graph_group_sync.cpp:302: undefined reference to `marian::ModelQuantizer::quantize(std::shared_ptr<marian::ExpressionGraph>)'
collect2: error: ld returned 1 exit status
3rd_party/marian-dev/src/CMakeFiles/marian_train.dir/build.make:104: recipe for target 'marian' failed
make[2]: *** [marian] Error 1
CMakeFiles/Makefile2:729: recipe for target '3rd_party/marian-dev/src/CMakeFiles/marian_train.dir/all' failed
make[1]: *** [3rd_party/marian-dev/src/CMakeFiles/marian_train.dir/all] Error 2
Makefile:155: recipe for target 'all' failed
make: *** [all] Error 2
Build step 'Execute shell' marked build as failure
Finished: FAILURE
```

This is [browsermt/marian-dev@wasm](https://github.com/browsermt/marian-dev/tree/wasm) fail. I don't want to overwrite the marian-dev here with the master (which works, I've tested locally) since WASM dependent on it.

**Comment by jerinphilip** [comment]
@abhi-agg: Since you're already on one switch for WASM vs NATIVE, can I ask you to take up moving PCRE2 to the system one an not the 3rd-party/.. one in native builds? This is an adjustment of the associated CMake option, since you're editing and working on merge here I'd rather not interfere until stable.

**Comment by abhi-agg** [comment]
Will submit a fix for both soon 👍

File: CMakeLists.txt
Comment (abhi-agg) [reply CMakeLists.txt]: [This commit](https://github.com/browsermt/bergamot-translator/pull/8/commits/c2b1c6eab484de4acf8d19fb334998927567978e) fixes this.

**Comment by abhi-agg** [comment]
@jerinphilip Both of your issues should be resolved by last 2 commits. Please check at your end 👍

**Comment by jerinphilip** [comment]
@abhi-agg: I will confim browsermt/marian-dev@wasm's issues looks to be fixed with the recent commits. 
http://vali.inf.ed.ac.uk/jenkins/job/bergamot-translator/43/console

Until the time browsermt/marian-dev@wasm gets reviewed and merged to master (ensuring the disabled switches only disable them for wasm) by @kpu or someone else who's qualified (I'm probably not, at least for now) I will configure my jenkins builds with the following conf:

<details> <summary> Click to expand </summary>
<p>

```bash
. /etc/environment
rm -rf build
mkdir -p build
cd build
#CMAKE=/usr/bin/cmake
CMAKE=/opt/modules/cmake-3.20-rc1/bin/cmake
CC=/usr/bin/gcc-8 CXX=/usr/bin/g++-8 \
	$CMAKE \
	-DCOMPILE_CUDA=off -DCOMPILE_TESTS=OFF -DCOMPILE_EXAMPLES=OFF  \
        -DUSE_SENTENCEPIECE=ON -DCOMPILE_SERVER=OFF -DUSE_WASM_COMPATIBLE_MARIAN=off \
    ..

# Build time patch: browsermt/marian-dev@wasm -> browsermt/marian-dev@master
git -C $WORKSPACE/3rd_party/marian-dev fetch -a 
git -C $WORKSPACE/3rd_party/marian-dev checkout origin/master
git -C $WORKSPACE/3rd_party/marian-dev rev-parse --short HEAD
git -C $WORKSPACE add 3rd_party/marian-dev
( git config --global user.name 'Marian Minion';
git config --global user.email 'jenkins@vali.inf.ed.ac.uk';
git -C $WORKSPACE commit \
	-m "Jenkins: Local submodule update"
)
# Display CMake cached variables for debugging
$CMAKE -L -N
make -j8
# No tests for bergamot-translator
# CUDA_VISIBLE_DEVICES=0 env CTEST_OUTPUT_ON_FAILURE=1 make test
cd ..
# Create the artifact with files needed for regression tests
tar zcvf bergamot-translator.tgz \
	build/CMakeCache.txt \
    build/libmarian.a build/libssplit.a build/src/translator/libbergamot-translator.a \
    build/marian*  build/spm_*  \
    build/app/* 
```
</p>
</details>

This should let us know if something in wasm branch breaks stuff here again. GitHub CI workflows file (the original ubuntu, macos) for vanilla configuration should only require the addition of `-DUSE_WASM_COMPATIBLE_MARIAN=off` if all works out well, it'd faster if you check those in with the modifications. 

There also seems to be a segfault at full WNGT20 decode (bergamot-translator-regression-tests) are [failing](http://vali.inf.ed.ac.uk/jenkins/job/bergamot-translator-regression-tests/31/). Hold off on merge to main until I isolate the cause.

**Comment by abhi-agg** [comment]
> I will configure my jenkins builds with the following conf:

You mean, this config change will happen only for
- [ ] native build with vanilla marian (no WASM modifications, should build against all marian-capabilities)

And NOT for these 2 builds:
- [ ] native build with marian-modifications for WASM @abhi-agg 
- [ ] wasm build with marian-modifications for WASM @abhi-agg 

Right?

**Comment by jerinphilip** [comment]
@abhi-agg GitHub CI will have the 3, no modifications. It's not a change, the 3 builds will work with the source here. I'll have a version (separately) compiling with master branch from marian-dev, that's what the comment means. If something in browsermt/marian-dev@wasm breaks (like the quantizer), we'll know.

**Comment by abhi-agg** [comment]
@jerinphilip
- [ ] native build with vanilla marian (no WASM modifications, should build against all marian-capabilities) @jerinphilip 
- [x] native build with marian-modifications for WASM @abhi-agg 
- [x] wasm build with marian-modifications for WASM @abhi-agg 

Last 2 are done. Once you push the change for the 1st one, all 3 should be good to go.

File: src/translator/translation_result.h
Comment (jerinphilip) [reply src/translator/translation_result.h]: These are no longer public. Marking as resolved.

**Review by abhi-agg — APPROVED** [review]
> bergamot-translator is building: 
> * http://vali.inf.ed.ac.uk/jenkins/job/bergamot-translator/48/
> 
> > There also seems to be a segfault at full WNGT20 decode (bergamot-translator-regression-tests) are failing. Hold off on merge to main until I isolate the cause.
> 
> 
> No issues at bergamot-translator-tests:
> * http://vali.inf.ed.ac.uk/jenkins/job/bergamot-translator-regression-tests/35/
> 
> Automation is setup for both to open issues at respective repositories from jenkins@vali builds.

Lgtm. As we agreed, we will add tests for the artifacts built with wasm compatible sources in bergamot-translator-tests repo.

**Comment by jerinphilip** [comment]
This has become one gigantic PR. Mozilla has branched this (integration) into wasm-integration and tinkered with threads, meanwhile I have made more refactors and adjustments in an attempt to fit to WASM development in a cleaner way. Everyone seems to be interested in having a 'main' as well.

Here's a summary of changes since @kpu's last review, for doc:

1. Two TranslationResults are maintained - as marian specific `marian::bergamot::Response` and  translator agnostic `TranslationResult`. This should simplify development for people involved despite the overhead of conversion. 
2. `BatchTranslator` has been refactored for a cleaner path (https://github.com/browsermt/bergamot-translator/issues/10). `BatchTranslator` and `Batcher` are no longer aware of threads. Threads are completely in `Service`. This is eventually to become a replacement marian-server/decoder which can handle requests instead of files. 
3. `Service` is split as a `ServiceBase` from which `Service` and `NonThreadedService` inherits from. `ServiceBase` keeps the code common to both, specifies unimplemented `enqueue()` and `stop()`, which is implemented differently in the threaded and non-threaded version eliminating redundancies and `#ifdef` troubles. Seems to cover https://github.com/browsermt/bergamot-translator/issues/41.
4. GitHub CI is setup for both wasm, native with wasm adjustments and vanilla (which is native with full marian). Builds are passing.
5. Jenkins is setup with bergamot-translator build time swapped with browsermt/marian-dev@master instead of browsermt/marian-dev@wasm to identify wasm modifications breaking anything. With this I'm okay having unreviewed browsermt/marian-dev@wasm in this source.
6. Jenkins is setup with bergamot-translator-regression-tests for functionality of single-thread, multi-thread and an intensive case of WNGT20 to ensure speed is retained and potential deadlocks/segfaults show up. 

Note for @kpu: Concurrent queuing (https://github.com/browsermt/bergamot-translator/issues/41) is not included in this PR.
I'm not requesting a further review (which is cumbersome given how big this has become). I request to carry forward any concerns in respective issues. With the current state of source, this PR will close https://github.com/browsermt/bergamot-translator/issues/41, https://github.com/browsermt/bergamot-translator/issues/10, https://github.com/browsermt/bergamot-translator/issues/27, https://github.com/browsermt/bergamot-translator/issues/11.

@abhi-agg please merge into a 'main'. I understand risks and take responsibility for edits/ potential issues from the unreviewed code in my edits.

**Comment by abhi-agg** [comment]
Cool. Merging it.

**Pull request merged**
